### PR TITLE
WT-11603 Call direct pack functions rather than run-time format parsing

### PR DESCRIPTION
**Pull request by ershov**
* `dist/log.py` now generates packing and unpacking functions so that they directly call lower level routines rather than by parsing the format at run time from a C string.
* Since tests are using both old and new code and the old code is not in the repo anymore, the tests for this are in the [mongo-tests](https://github.com/wiredtiger/mongo-tests) repo in a corresponding subdirectory.

### CODE REVIEW

**Comment by ershov** [comment]
Please ignore the `s_all` and `clang analyser` errors for now. They are related to recent [WT-12144](https://jira.mongodb.org/browse/WT-12144). It will be fixed with a rebase before merge.

**Comment by agorrod** [comment]
I'm going to dip out of this. There are enough reviewers with more expertise in the specific area than me.

File: test/packing/log_auto_test.h
Comment (korteland) [inline test/packing/log_auto_test.h:47]: Didn't go over the entirety of this file either, but a couple of things:
* Some commented-out code
* A couple of comments that don't align with the style guide, i.e. start with uppercase and end with a full stop
* Would it be possible to reuse some existing test infra like `testutil_assert`?

**Review by korteland — COMMENTED** [review]
Wow - great work so far. I focused on the details rather than the overall shape (as discussed offline).

File: dist/log.py
Comment (sueloverso) [inline dist/log.py:95]: Why do all of these have double underscores in the name? Please just use one underscore except at the beginning of the name.

File: dist/log.py
Comment (sueloverso) [reply dist/log.py:334]: "PEEK the operation type" is not a sentence. If you really want to change it, it should be "Peek at ..."

File: test/packing/log_auto-test-gen.sh
Comment (sueloverso) [inline test/packing/log_auto-test-gen.sh:1]: Can we not have file names with mixed underscores and dashes please?

File: test/packing/log_auto-test-gen.sh
Comment (sueloverso) [reply test/packing/log_auto-test-gen.sh:15]: I think all of this needs a lot more commenting about what it is doing in general.

File: test/packing/packing-direct-test-all.c
Comment (sueloverso) [inline test/packing/packing-direct-test-all.c:55]: Can all of these TODOs actually have a comment describing the functions?

**Review by kommiharibabu — APPROVED** [review]
LGTM after resolving the review comments that are pending

File: test/packing/log_auto-test-gen.sh
Comment (ershov) [reply test/packing/log_auto-test-gen.sh:22]: Because not all functions in the generated code are used. The functions are generated based on the meta-data, it's not easy to select which functions to drop.

File: dist/log.py
Comment (ershov) [reply dist/log.py:334]: Yes, the formatting is fixed by `s_clang_format` afterwards.

Comment (ershov) [reply dist/log.py:334]: The word "Read" implies that the pointer advances when the function is returned. What actually happens is - the function "previews" the value but the text read operation will read it again.
The best would be to change the function name, this needs to be a separate work.

Capitals were supposed to highlight that special behavior but that might be an incorrect method of expression.

File: dist/s_void
Comment (ershov) [reply dist/s_void:172]: `s_void` is buggy and needs to be re-worked. It gives a lot of false positives.

File: dist/log.py
Comment (ershov) [reply dist/log.py:257]: That's how I tried to separate a suffix (decoded type name) and the main part (structure name).

File: dist/log.py
Comment (ershov) [reply dist/log.py:78]: Done

File: dist/log.py
Comment (ershov) [reply dist/log.py:261]: {TODO}

File: dist/log.py
Comment (ershov) [reply dist/log.py:92]: Done

File: dist/log.py
Comment (ershov) [reply dist/log.py:622]: In python, double newlines separate global definitions.

File: src/log/log_auto.c
Comment (ershov) [reply src/log/log_auto.c:12]: Fixed.

File: test/packing/CMakeLists.txt
Comment (ershov) [reply test/packing/CMakeLists.txt:10]: `wd` disables a warning. This one is for unknown pragmas.

File: test/packing/log_auto-test-gen.sh
Comment (ershov) [reply test/packing/log_auto-test-gen.sh:15]: I have added more comments.
The scripts themselves are going into another repo - `mongo-tests`.

File: test/packing/log_auto-test-gen.sh
Comment (ershov) [reply test/packing/log_auto-test-gen.sh:45]: Done

File: dist/log.py
Comment (ershov) [reply dist/log.py:95]: Double underscores are following the existing naming convention for similar functions. There is a ticket to fix it in the entire project: WT-11788.

File: test/packing/packing-direct-test-all.c
Comment (ershov) [reply test/packing/packing-direct-test-all.c:55]: Done

File: dist/log.py
Comment (ershov) [reply dist/log.py:360]: {TODO}

File: dist/s_void
Comment (ershov) [reply dist/s_void:172]: WT-12355

File: dist/log.py
Comment (korteland) [reply dist/log.py]: This is still outstanding - I'd prefer for the post-formatted code to match the generated code. (My reasoning here is that (1) it's easier to grep for code between the generator script and generated C, and (2) running the script would ideally leave the checked out in a fairly clean state, without forcing us to fix up the whitespace afterwards in a separate step.)

Comment (korteland) [reply dist/log.py]: Thanks for fixing this one - closing the thread out.

File: src/log/log_auto.c
Comment (korteland) [reply src/log/log_auto.c:18]: Thanks!

File: dist/log.py
Comment (korteland) [reply dist/log.py:289]: I think I can resolve this, but I need to check. GitHub doesn't say the comment is outdated, and the code no longer seems to have macros relying on specific variable names. But I definitely remember there were such macros.

Is this something you fixed?

Comment (ershov) [reply dist/log.py:289]: I fixed all instances that I could find. GitHub doesn't indicate that because this particular line hasn't changed, I guess :)

File: dist/log.py
Comment (ershov) [reply dist/log.py]: I think, the reasonable solution is somewhere in between: keep the generated code readable but not aim to exactly fit the current formatting rules. The reasons for that are:
1. The generated code must not be cryptic.
2. Formatting standards can change, we don't want to change every generator then.
3. Strictly following formatting rules unnecessarily clutters the generator's logic.

Comment (korteland) [reply dist/log.py]: That makes sense, and I agree for any functions that are generated "piecewise". But for this particular section, it looks like we have entire functions embedded in the script with no templating. I think these could easily be reformatted to match the generated form.

Comment (ershov) [reply dist/log.py]: Fixed formatting for most varbatim fragments where "proper" formatting doesn't obscure the code.

**Review by korteland — APPROVED** [review]
LGTM - thanks for getting this through. It's been a long road and I appreciate your patience.

**Review by sueloverso — APPROVED** [review]
LGTM. I have one question. Have you confirmed these are in-place changes and logs are readable and recoverable across old/new versions? That is the only remaining general concern I have.

**Comment by ershov** [comment]
> Have you confirmed these are in-place changes and logs are readable and recoverable across old/new versions? That is the only remaining general concern I have.

Yes, one of the tests is a "mirror test". What it does is - replacing every packing and unpacking function with one which calls both variants and compares their outputs. With this modification I ran every test that we have in WT including the ones that read or write logs. I never got an assert fired. This means that every new packing and unpacking function produces result identical to the old version.

If you have any tests in mind that I could run outside of our WT tests, do let me know. I want it to be as reliable as it can be.

File: dist/log.py
Comment (kommiharibabu) [inline dist/log.py:218]: what is the need for the above two unpack check macros? Doesn't a single one satisfy the requirement?

File: dist/log.py
Comment (kommiharibabu) [inline dist/log.py:530]: can you add a comment at the top of this file to explain the use of these macros? Also, could you mention that these will be removed in future versions? I do not find any need for these macros other than testing, but we can keep them for one major release and remove them.

File: dist/log.py
Comment (ershov) [reply dist/log.py:218]: The ">=" versions check that there is more space in the buffer (i.e. at least one byte more left), the ">" one checks that the pointer has not gone beyond the end of the buffer (i.e. allows that 0 bytes are left).

File: dist/log.py
Comment (ershov) [reply dist/log.py:539]: Done

File: dist/log.py
Comment (kommiharibabu) [reply dist/log.py:218]: Thanks.

**Comment by ershov** [comment]
It looks to me like this PR is ready to merge. I'll give it one more day for follow-up and submit it tomorrow if there are no objections.

**Pull request merged**
